vss_dedupe.txt

```diff
@@ -0,0 +1,572 @@
+  sensor    string LowVoltageSystemState
+  sensor    float Speed
+  sensor    float TravelledDistance
+  sensor    float TraveledDistance
+  sensor    float TraveledDistanceSinceStart
+  sensor    float TripDuration
+  sensor    boolean IsBrokenDown
+  sensor    boolean IsMoving
+  sensor    float AverageSpeed
+  sensor    uint16 CurrentOverallWeight
+  actuator  float TripMeterReading
+  actuator  uint8 PowerOptimizeLevel
+['Vehicle:LowVoltageBattery']
+  attribute uint16 NominalVoltage
+  attribute uint16 NominalCapacity
+  sensor    float CurrentVoltage
+  sensor    float CurrentCurrent
+['Vehicle:Acceleration']
+  sensor    float Longitudinal
+  sensor    float Lateral
+  sensor    float Vertical
+['Vehicle:AngularVelocity']
+  sensor    float Roll
+  sensor    float Pitch
+  sensor    float Yaw
+['Vehicle:Trailer']
+  sensor    boolean IsConnected
+['Vehicle:CurrentLocation']
+  sensor    string Timestamp
+  sensor    double Latitude
+  sensor    double Longitude
+  sensor    double Heading
+  sensor    double HorizontalAccuracy
+  sensor    double Altitude
+  sensor    double VerticalAccuracy
+['Vehicle:CurrentLocation:GNSSReceiver']
+  sensor    string FixType
+['Vehicle:CurrentLocation:GNSSReceiver:MountingPosition']
+  attribute int16 X
+  attribute int16 Y
+  attribute int16 Z
+['Vehicle:Powertrain']
+  attribute string Type
+  sensor    float AccumulatedBrakingEnergy
+  sensor    uint32 Range
+  actuator  uint8 PowerOptimizeLevel
+['Vehicle:Powertrain:CombustionEngine']
+  attribute string EngineCode
+  attribute uint16 Displacement
+  attribute float StrokeLength
+  attribute float Bore
+  attribute string Configuration
+  attribute uint16 NumberOfCylinders
+  attribute uint16 NumberOfValvesPerCylinder
+  attribute string CompressionRatio
+  attribute float EngineOilCapacity
+  attribute float EngineCoolantCapacity
+  attribute uint16 MaxPower
+  attribute uint16 MaxTorque
+  attribute string AspirationType
+  sensor    string EngineOilLevel
+  sensor    int32 OilLifeRemaining
+  sensor    boolean IsRunning
+  sensor    uint16 Speed
+  sensor    float EngineHours
+  sensor    float IdleHours
+  sensor    int16 ECT
+  sensor    int16 EOT
+  sensor    uint16 _MAP
+  sensor    uint16 MAF
+  sensor    uint8 TPS
+  sensor    uint16 EOP
+  sensor    uint16 Power
+  sensor    uint16 Torque
+['Vehicle:Powertrain:CombustionEngine:DieselExhaustFluid']
+  attribute float Capacity
+  sensor    uint8 Level
+  sensor    uint32 Range
+  sensor    boolean IsLevelLow
+['Vehicle:Powertrain:CombustionEngine:DieselParticulateFilter']
+  sensor    float InletTemperature
+  sensor    float OutletTemperature
+  sensor    float DeltaPressure
+['Vehicle:Powertrain:Transmission']
+  attribute string Type
+  attribute int8 GearCount
+  attribute string DriveType
+  sensor    float TravelledDistance
+  sensor    int8 CurrentGear
+  sensor    int16 Temperature
+  sensor    uint8 ClutchWear
+  actuator  int8 SelectedGear
+  actuator  boolean IsParkLockEngaged

```

Comment:
One issue that I see is that the VSS entries are in arbitrary/logical order in order to flow more logically and be more human readable, but given the variability in data types, this is not conducive to optimal struct layout. In this case, there are mixed 8/16/32bit entries, where if we were to take this with a 1:1 mapping, any dereference of DriveType on a non-byte addressable system will generate an alignment trap. The in-memory size of the struct could also change depending on whether the compiler automatically inserts padding bytes or not, which could be an issue if there is an intention to do binary serialization/deserialization of the structure via things like gRPC across systems with different memory models. While this could be mitigated by forcing the compiler to pack the struct and leave it as-is, this would again go back to triggering alignment traps on non-aligned dereferencing.

One possible solution would be to have an intermediate step in which the generated items are passed through an analysis tool that reorders struct elements and adds implicit padding bytes where needed to match alignment needs.

Reply:
For DDS (at least), the details of byte-addressability, endianness, word size, etc. are handled by the DDS implementation for that platform.  Systems with different endianness/CPU/OS/ProgLang will interoperate without knowing what's on the other end(s) of the DDS connection.
But this is a good point -- the order of appearance in a struct is implied by where the element is placed in the .vspec file, not by something explicit (such as a keyword that states the position of this element in a struct).
Is implicit order good enough?